### PR TITLE
[FLINK-5706] [file systems] Add S3 file systems without Hadoop dependencies

### PR DESCRIPTION
**Pull request by StephanEwen**
## What is the purpose of the change

This adds two implementations of a file system that write to S3 so that users can use Flink with S3 without depending on Hadoop and have an alternative to Hadoop's S3 connectors.

Both are not actual re-implementations but wrap other implementations and shade dependencies.

1. The first is a wrapper around Hadoop's s3a file system. By pulling a smaller dependency tree and shading all dependencies away, this keeps the appearance of Flink being Hadoop-free, from a dependency perspective. We can also bump the shaded Hadoop dependency here to get improvements to s3a in (as in Hadoop 3.0) without causing dependency conflicts.

2. The second S3 file system is from the Presto Project. Initial simple tests seem to indicate that it responds slightly faster and in a bit more lightweight manner to write/read/list requests, compared to the Hadoop s3a FS, but it has some semantical differences. For example, creating a directory does not mean the file system recognized that the directory is there. The directory is only recognized as existing once files are inserted. For checkpointing, that could even be preferable.

Both file systems register themselves under `s3` to not overload the `s3n` and `s3a` schemes used by Hadoop.

Both file systems are not in the classpath of Flink by default, but must be explicitly added. They register themselves via the service mechanism. Their jars are built into the `/opt` directory in the distribution.

## Brief change log

  - Adds `flink-filesystems/flink-s3-fs-hadoop`
  - Adds `flink-filesystems/flink-s3-fs-presto`

## Verifying this change

This adds some initial integration tests, which do depend on S3 credentials. These credentials are not in the code, but only encrypted on Travis, which is why the tests can only run in a meaningful way either on the `apache/flink` master branch, or in a committer repository when the committer enabled Travis uploads to S3 (for logs) - the tests here use the same secret credentials.

Since this does not implement the actual S3 communication, we have no tests for that. The tests only test instantiation and whether S3 communication can be established (simple reads/writes to a bucket, listing, etc).

Change can also be verified by building Flink, pulling the respective S3 FS jar from `/opt` into `/lib` and running a workload that checkpoints or writes to S3.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)

Proper behavior of the File Systems is important, otherwise checkpointing may fail. In some sense we are already relying on proper tests of HDFS and S3 connectors by the Hadoop project. This adds a similar dependency.

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

Will add documentation once the details of this feature are agreed upon.

### CODE REVIEW

File: flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/S3FileSystemFactory.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	private static final Logger LOG = LoggerFactory.getLogger(S3FileSystemFactory.class);
+
+	/** The prefixes that Flink adds to the Hadoop config under 'fs.s3a.'. */
+	private static final String[] CONFIG_PREFIXES = { "s3.", "s3a.", "fs.s3a." };
+
+	/** Keys that are replaced (after prefix replacement, to give a more uniform experience
+	 * across different file system implementations. */
+	private static final String[][] MIRRORED_CONFIG_KEYS = {
+			{ "fs.s3a.access-key", "fs.s3a.access.key" },
+			{ "fs.s3a.secret-key", "fs.s3a.secret.key" }
```
Comment (steveloughran) [inline flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/S3FileSystemFactory.java]: think about the session key; lets you support temporary credentials

**Comment by steveloughran** [comment]
1. I hope you pick up Hadoop 2.8.1 for this, as it's got a lot of the optimisations
1. And equally importantly: a later SDK
1. Though not one of the more recent 1.11 SDKs, where support is yet to ship. That's a big single shaded aws-sdk JAR so things like joda-time, jackson, guava, etc, don't cause problems, just the detail of 50+MB more of .jar on the CP.

Test wise, see how well your client works with a v4 endpoint like frankfurt, as there you also have to change the endpoint used.

Otherwise, nothing obvious I'd flag up as dangerous

File: flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/S3FileSystemFactory.java (hunk 1 above)
Comment (StephanEwen) [reply flink-filesystems/flink-s3-fs-hadoop/src/main/java/org/apache/flink/fs/s3hadoop/S3FileSystemFactory.java]: Thanks, I think we will probably add more keys here over time.

**Comment by StephanEwen** [comment]
Thanks @steveloughran for the comments.

I am actually using Hadoop 2.8.1 here with AWS SDK 1.11.95.
The shaded artifacts are only a few MBs large, so this seems okay.

**Pull request closed without merging**